wikimedia/wikipedia-ios
# Save additional notification fields in RemoteNotifications database

**Phabricator:** https://phabricator.wikimedia.org/T287298

### Notes
Building off of work in https://github.com/wikimedia/wikipedia-ios/pull/4018, this PR captures as much data as I think we will need from the notifications objects returned from the API. This also updates fields that do not need to be optional. A couple of callouts worth noting:

1. Previously we ignored the server-side `read` flag and maintained a local read / seen / excluded state. Now we are simplifying into a single `isRead` flag that is powered off of the API value. I think it's good to keep this simple for now, and add back in a local read / seen / excluded flag later if we find we need it.
2. This also stores the links dictionary from the notifications object as a transformable field in the database. To me this seemed simpler than maintaining a complicated separate set of links managed objects with primary and secondary relationships. It decodes into a `RemoteNotificationsLinks` object.

After review, please wait for #4018 to be reviewed and merged. Then update this with `feature/notifications` and merge.

It would probably help to add the code changes mentioned in https://github.com/wikimedia/wikipedia-ios/pull/4018 to allow for more frequent polling and looser API parameters.

### Test Steps
1. Launch on freshly installed app. Wait for notifications polling to occur.
2. Inspect database, confirm the records added looks right.
3. Terminate app. Add `let messageLinks = readNotifications.first?.messageLinks` right before the `markAsRead` call to the `apiController` in `RemoteNotificationsMarkAsReadOperation` [here](https://github.com/wikimedia/wikipedia-ios/blob/T287298-2/WMF%20Framework/Remote%20Notifications/Operations/RemoteNotificationsMarkAsReadOperation.swift#L9). Also add a breakpoint here. Launch app. When breakpoint is hit, confirm `messageLinks` value looks right.
